Add streaming graph persistence layer from kafka to file

tmkasun opened the pull request:
In this PR:

- Add `JasmineGraphInstanceProtocol` command to sent streaming graph edges from `frontendservicesesion` to worker's `instanceservicesession`
![image](https://user-images.githubusercontent.com/3313885/112755689-73f80a80-8fff-11eb-8f23-7b4f6fcbae98.png)

miyurud commented on src/centralstore/incremental/DataPublisher.cpp:
```diff
@@ -0,0 +1,53 @@
+}
+
+void DataPublisher::publish(std::string message) {
+    char recever_buffer[1024] = {0};
```
Lets declare the 1024 as a constant. See for [example](https://github.com/miyurud/jasminegraph/blob/master/src/frontend/JasmineGraphFrontEndProtocol.h#L85).

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/DataPublisher.cpp:
```diff
@@ -0,0 +1,53 @@
+    // Receve ack for edge data content length
+
+    if (return_status > 0) {
+        fprintf(stdout, "Received int = %d\n", ntohl(received_int));
```
I do not see any reason for having this `fprintf()` here. The message is also not intuitive.

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud commented on src/centralstore/incremental/DataPublisher.h:
```diff
@@ -0,0 +1,21 @@
+#include <arpa/inet.h>
```
Add copyright header

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud replied:
Done

miyurud commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -221,6 +223,11 @@ NodeBlock* NodeBlock::get(unsigned int blockAddress) {
         node_block_logger.error("Error while reading edge reference data from block " + std::to_string(blockAddress));
```
Don't we have to say which graph ID and which partition ID we are referring to here?

tmkasun replied:
I wish we could, But in this class, we don't have graph & partition information, Graph ID and Partition ID is used to create the NodeDB database file so it's in the file name, But in this class, all we have is the file stream [`std::fstream*`](https://github.com/miyurud/jasminegraph/blob/740e27dd015904c4fd14483546dcedafc25686e9/src/centralstore/incremental/NodeBlock.cpp#L264), to the DB file , But it seems we can't get the file path from `fstream` or at least it's not straightforward

- [SO Answer 1](https://stackoverflow.com/questions/3022692/how-get-file-path-from-stdifstream-in-c)
- [SO Answer 2](https://stackoverflow.com/questions/10773391/getting-filename-or-path-from-fstream)

miyurud replied:
Based on the discussion we had today, it seems that it is not possible to link the graph ID and partition ID from the execution context. Hence we will keep this error message as it is and proceed.

miyurud commented on src/centralstore/incremental/NodeBlock.cpp:
```diff
@@ -221,6 +223,11 @@ NodeBlock* NodeBlock::get(unsigned int blockAddress) {
     }
 
+    if (!NodeBlock::nodesDB->read(reinterpret_cast<char*>(&edgeRefPID), sizeof(unsigned char))) {
+        node_block_logger.error("Error while reading edge reference partition ID data from block " +
```
Don't we have to say which graph ID and which partition ID we are referring to here?

tmkasun replied:
Not at this NodeDB class level as mentioned in https://github.com/miyurud/jasminegraph/pull/122#discussion_r606796318

miyurud replied:
Ok

miyurud commented on src/frontend/JasmineGraphFrontEnd.cpp:
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 #include <map>
 #include <set>
 #include "JasmineGraphFrontEnd.h"
+#include <nlohmann/json.hpp>
```
This is a new library for JasmineGraph. We need to know the license and also update the information in the README.

tmkasun replied:
This has [MIT license](https://github.com/nlohmann/json/blob/develop/LICENSE.MIT), Will update the README file as well

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud replied:
Ok

miyurud commented on src/frontend/JasmineGraphFrontEnd.cpp:
```diff
@@ -50,6 +53,7 @@ void *frontendservicesesion(std::string masterIP, int connFd, SQLiteDBInterface
     frontend_logger.log("Master IP: " + masterIP, "info");
     char data[FRONTEND_DATA_LENGTH];
     bzero(data, FRONTEND_DATA_LENGTH + 1);
+    DataPublisher workerClient(7780, "127.0.0.1");
```
Why is this host IP and port hard coded here?

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/7174d9c7814f01e19de59293f8917ecd669a662f

miyurud replied:
Ok, done

miyurud commented on src/frontend/JasmineGraphFrontEnd.cpp:
```diff
@@ -515,7 +518,17 @@ void *frontendservicesesion(std::string masterIP, int connFd, SQLiteDBInterface
+                std::string sId = std::string(sourceJson["id"]);
+                std::string dId = std::string(destinationJson["id"]);
+
+                partitionedEdge pe = graphPartitioner.addEdge({sId, dId});
```
Lets have better variable name than `pe`

miyurud replied:
Done

miyurud commented on src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp:
```diff
@@ -0,0 +1,29 @@
+#include <memory>
```
Missing license header

miyurud replied:
Done

miyurud commented on src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp:
```diff
@@ -0,0 +1,29 @@
+    try {
+        auto edge = json::parse(edgeString);
+        NodeBlock *sourceNode = this->nm->addNode(edge["id"]);
+        // NodeBlock* destinationNode = nodeManagers.at(destinationJson["pid"].get<int>())->addNode(dId);
```
Remove the commented code

miyurud replied:
Done

miyurud commented on src/localstore/incremental/JasmineGraphIncrementalLocalStore.cpp:
```diff
@@ -0,0 +1,29 @@
+        // NodeBlock* destinationNode = nodeManagers.at(destinationJson["pid"].get<int>())->addNode(dId);
+    } catch (const std::exception &) {
+        incremental_localstore_logger.log("Error malformed JSON attributes!", "error");
+        // TODO handle JSON errors
```
When you type a TODO please use the format ``TODO<github-user-name>: message``

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud replied:
Done

miyurud commented on src/localstore/incremental/JasmineGraphIncrementalLocalStore.h:
```diff
@@ -0,0 +1,16 @@
+#include <string>
```
Add license header

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud replied:
Done

miyurud commented on src/partitioner/stream/Partition.cpp:
```diff
@@ -0,0 +1,152 @@
+
+float Partition::edgeCutsRatio() { return this->edgeCutsCount() / (this->getEdgesCount() + this->edgeCutsCount()); }
+
+void Partition::printEdgeCuts() {
```
This function seems to be for debugging purpose. Shall we make the `cout` statements to debug log lines?

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud replied:
Done

miyurud commented on src/partitioner/stream/Partition.cpp:
```diff
@@ -0,0 +1,152 @@
+    }
+}
+
+void Partition::printEdges() {
```
The same comment on `cout` applies here too.

tmkasun replied:
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud replied:
Done

miyurud commented on src/partitioner/stream/Partitioner.cpp:
```diff
@@ -209,8 +210,8 @@ partitionedEdge Partitioner::fennelPartitioning(std::pair<int, int> edge) {
-    // std::cout << "Vertext 1 = " << stoi(v[0]) << std::endl;
-    // std::cout << "Vertext 2 = " << stoi(v[1]) << std::endl;
+    std::vector<std::string> v = Partition::_split(data, ' ');
+    std::cout << "Vertext 1 = " << stoi(v[0]) << std::endl;
```
Why do we need these couts?

tmkasun replied:
These are for debugging purpose, Hence converted them to debug logs
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud replied:
Ok, Done

tmkasun commented on src/centralstore/incremental/DataPublisher.cpp:
```diff
@@ -0,0 +1,53 @@
+#include "./DataPublisher.h"
```
Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/bdd734e88390169ddc9e44c74030d1c1b170d0ae

miyurud replied:
Done

tmkasun commented:
> The use of the variable `JASMINE_HOME` here is ambiguous. Shouldn't this be just user's home directory?

As per the offline discussion , Renamed the `JASMINE_HOME` environment variable name to `JASMINEGRAPH_HOME` to prevent conflicting with default user `HOME` environment variable

Fixed in https://github.com/miyurud/jasminegraph/pull/122/commits/72bf122c4e928dfcf8739b15f2c51f67012a5c21

miyurud commented:
> > The use of the variable `JASMINE_HOME` here is ambiguous. Shouldn't this be just user's home directory?
> 
> As per the offline discussion , Renamed the `JASMINE_HOME` environment variable name to `JASMINEGRAPH_HOME` to prevent conflicting with default user `HOME` environment variable
> 
> Fixed in [72bf122](https://github.com/miyurud/jasminegraph/commit/72bf122c4e928dfcf8739b15f2c51f67012a5c21)

Ok, but we should keep `HOME` variable as it is. See the comment I have put there.
